### PR TITLE
ARROW-7008: [C++/Python] value_offsets may be omitted for 0-len arrays

### PR DESCRIPTION
**Pull request by xhochy**

### CODE REVIEW

File: cpp/src/arrow/array.cc [hunk 1]
```diff
@@ -1322,6 +1322,10 @@ struct ValidateVisitor {
     if (array.length() && !value_offsets) {
       return Status::Invalid("value_offsets_ was null");
     }
+    if (array.length() == 0 && !value_offsets) {
+      // In the case of an empty array value_offset may be omitted.
+      return Status::OK();
+    }
```
Comment (bkietz) [inline cpp/src/arrow/array.cc]: Nit: consolidate related conditions, don't rely on pointer to boolean conversion (especially if using `auto` to declare the pointer)
```suggestion
    if (value_offsets == nullptr) {
      return array.length() == 0
        // In the case of an empty array value_offset may be omitted.
        ? Status::OK()
        : Status::Invalid("value_offsets_ was null");
    }
```

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by xhochy** [comment]
Failure seems unrelated.

**Comment by wesm** [comment]
Interesting edge case from the reproducer in the JIRA. It isn't clear whether a null buffer or a length-1 value-offsets buffer should be required. Since the buffer should never be used I don't see an issue with it being omitted, but we probably would need to clarify that in Columnar.rst

@emkornfield or @jacques-n thoughts about this?

**Comment by emkornfield** [comment]
@wesm agreed on both accounts (should be OK for it to be missing and we should clarify it)

**Comment by pitrou** [comment]
The problem with null buffers is that it pushes the burden of checking for nulls on the consumer. It is allowed for null bitmaps when there are no nulls, as a space optimization, but IMHO allowing null buffers on 0-length arrays is a nuisance.

So I think the producer should be fixed, instead of relaxing the validation check.

**Comment by wesm** [comment]
Yes, I think we should fix the root cause of the original failure (the invalid array) and then we can discuss more about the null offsets buffer.

**Comment by emkornfield** [comment]
@xhochy what are the next steps here?

**Comment by wesm** [comment]
I'm going to close this and take a crack at fixing the immediate problem that `pa.array([], type=pa.string()).dictionary_encode().dictionary` produces an invalid array (that segfaults when you call `ValidateFull`). I'll open a follow up JIRA to address the matter of whether the offsets buffer can be null in a length 0 array. My intuition is that it should be OK to omit but we should deal with that separately from fixing the invalid array produced by that code

**Pull request closed without merging**